File: php/lib/jumpplus_downloader.php
<?php
/*
メモ
widthが760なら-6のずれ
widthが764なら-7のずれ
widthが822なら-5.5のずれ
*/
class jumppuls_downloader
{
    public function __construct()
    {
        $this->file = 0;
        $this->h = 1200;
        $this->w = 760;
        $this->h_marge = 4;
        $this->w_marge = 6;
    }


    public function auto_list_download($url, $next = false, $sleep = 0)
    {
        $this->json_download($url);
        if (!file_exists($this->list["readableProduct"]["title"])) {
            mkdir($this->list["readableProduct"]["title"], 0755);
        }
        foreach ($this->list["readableProduct"]["pageStructure"]["pages"] as $i => $page) {
            sleep($sleep);
            $this->file = $i;
            if ($page["type"] == "main") {
                $this->h = $page["height"];
                $this->w = $page["width"];
                $this->w_marge = $page["width"] / 4 % 8; //%の後の数字、8の倍数であることは間違いないが詳しくは不明 データ不足
                $this->download($page["src"]);
                $this->processing();
                $this->output("./" . $this->list["readableProduct"]["title"] . "/");
            }
        }
        if ($this->list["readableProduct"]["nextReadableProductUri"] != null and $next) {
            $this->auto_list_download($this->list["readableProduct"]["nextReadableProductUri"], true);
        }
    }

    public function json_download($url)
    {
        $jsondata = file_get_contents($url . ".json");
        $this->list = json_decode($jsondata, true);
    }

    public function download($url)
    {
        $this->img = imagecreatefromjpeg($url);
    }

    public function processing()
    {
        $this->dst_image = imagecreatetruecolor($this->w - $this->w_marge * 4, $this->h - $this->h_marge * 4);
        $width = $this->w / 4 - $this->w_marge;
        $height = $this->h / 4 - $this->h_marge;
        for ($i = 0; $i <= 3; $i++) {
            for ($ii = 0; $ii <= 3; $ii++) {
                imagecopyresampled($this->dst_image, $this->img, $width * $i, $height * $ii, $width * $ii, $height * $i, $width, $height, $width, $height);
            }
        }
    }
    public function output($file = "./")
    {
        imagepng($this->dst_image, $file . $this->file . '.png');
        $this->file++;
    }
}

File: py/lib/jumpplus_downloader.py
import glob
import html
import json
import math
import os
import re
import time
from io import BytesIO

import img2pdf
import requests
from PIL import Image


class jumpplus_downloader:
    def __init__(self, dir="./"):
        self.file = 0
        self.h = 1200
        self.w = 760
        self.session = requests.session()
        self.dir = dir

    def __get_headers(self):
        return {
            "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36"
        }

    def login(self, email_address, password, return_location_path="/"):
        data = {
            "email_address": email_address,
            "password": password,
            "return_location_path": return_location_path,
        }
        headers = {"x-requested-with": "XMLHttpRequest"}
        self.response = self.session.post(
            "https://shonenjumpplus.com/user_account/login",
            headers=dict(self.__get_headers(), **headers),
            data=data,
        )
        return self

    def auto_list_download(
        self, url, sleeptime=2, pdfConversion=True, zero_padding=True
    ):
        self.json_download(url)
        self.file = 0
        retitle = re.sub(
            r'[\\|/|:|?|.|"|<|>|\|]', "", self.list["readableProduct"]["title"]
        )
        if not os.path.isdir(self.dir + retitle):
            os.mkdir(self.dir + retitle)
        if self.list["readableProduct"]["pageStructure"]:
            for page in self.list["readableProduct"]["pageStructure"]["pages"]:
                time.sleep(sleeptime)
                if page["type"] == "main":
                    self.h = page["height"]
                    self.w = page["width"]
                    self.download(page["src"], False)
                    self.processing()
                    self.output(self.dir + retitle + "/", zero_padding=zero_padding)
            if pdfConversion:
                self.convertToPdf(self.dir + retitle + "/")

    def json_download(self, url):
        # Counterfeit User agent for absolutely successfully connection.
        text = self.session.get(url, headers=self.__get_headers()).text
        data = re.search(
            r"<script id='episode-json' type='text/json' data-value='(.*?)'>", text
        )
        self.list = json.loads(html.unescape(data.group(1)))

    def json_localread(self, filepath):
        with open(filepath) as json_file:
            json_data = json.load(json_file)
            self.list = json_data

    def download(self, url, fakeque=False):
        if fakeque:
            print("Emulating Download : " + url)
            self.img = url
        else:
            self.img = self.session.get(url, headers=self.__get_headers())

    def processing(self):
        readImage = Image.open(BytesIO(self.img.content))
        imageSize = readImage.size
        divideNum = 4
        multiple = 8
        width = math.floor(float(imageSize[0]) / (divideNum * multiple)) * multiple
        height = math.floor(float(imageSize[1]) / (divideNum * multiple)) * multiple
        buff = []
        counterX = 0
        counterY = 0

        for wx in range(4):
            inbuff = []
            for lx in range(4):
                cropped = readImage.crop(
                    box=(
                        width * counterX,
                        height * counterY,
                        width * (counterX + 1),
                        height * (counterY + 1),
                    )
                )
                inbuff.append(cropped)
                counterY += 1
            buff.append(inbuff)
            counterX += 1
            counterY = 0

        self.converted_img = readImage.copy()
        counterX = 0
        counterY = 0
        for wdx in buff:
            for ldx in wdx:
                self.converted_img.paste(
                    ldx, (int(width * counterX), int(height * counterY))
                )
                counterX += 1
            counterX = 0
            counterY += 1

    def output(self, dir, zero_padding=True):
        file = str(self.file)
        if zero_padding:
            index = 1
            zfill = 0
            while len(self.list["readableProduct"]["pageStructure"]["pages"]) >= index:
                index *= 10
                zfill += 1
            file = file.zfill(zfill)
        self.converted_img.save(dir + file + ".png")
        self.file += 1

    def convertToPdf(self, dir):
        img = [
            i.replace("\\", "/")
            for i in glob.glob(glob.escape(dir) + "*")
            if not i.endswith(".pdf")
        ]
        with open(dir + "output.pdf", "wb") as f:
            f.write(img2pdf.convert(img))

    # A simple Json Dumper for debugging.
    def dumpSimplifiedJson(self, jsObject):
        f = open("JSON.json", "w")
        json.dump(
            jsObject,
            f,
            ensure_ascii=False,
            indent=4,
            sort_keys=True,
            separators=(",", ": "),
        )
